File: decode_utf-8/main.zig
const std = @import("std");
const bitops = @import("bitops.zig");

const DecodeResult = struct {
    r: u32,
    s: usize,
    err: ?[]const u8,
};

/// Verifica se o ponto de código Unicode é válido. Retorna DecodeResult com erro, ou null se estiver ok.
fn validateRune(r: u32, b0: u8, b1: u8, expected_len: usize) ?DecodeResult {
    if (expected_len == 4) {
        if (r < 0x10000 or (b0 == 0xF0 and b1 < 0x90)) return DecodeResult{ .r = 0, .s = 0, .err = "overlong" };
        if (r >= 0xD800 and r <= 0xDFFF) return DecodeResult{ .r = 0, .s = 0, .err = "surrogate halfs" };
        if (r > 0x10FFFF) return DecodeResult{ .r = 0, .s = 0, .err = "too big" };
        return null; // OK
    }
    if (expected_len == 3) {
        if (r < 0x800 or (b0 == 0xE0 and b1 < 0xA0)) return DecodeResult{ .r = 0, .s = 0, .err = "overlong" };
        if (r >= 0xD800 and r <= 0xDFFF) return DecodeResult{ .r = 0, .s = 0, .err = "surrogate halfs" };
        if (r > 0x10FFFF) return DecodeResult{ .r = 0, .s = 0, .err = "too big" };
        return null; // OK
    }
    if (expected_len == 2 and r) {
        if (r < 0x80) return DecodeResult{ .r = 0, .s = 0, .err = "overlong" };
        if (r >= 0xD800 and r <= 0xDFFF) return DecodeResult{ .r = 0, .s = 0, .err = "surrogate halfs" };
        if (r > 0x10FFFF) return DecodeResult{ .r = 0, .s = 0, .err = "too big" };
        return null; // OK
    }

    return null; // OK
}

pub fn decodeRune(b: []const u8) DecodeResult {
    if (b.len == 0) {
        return DecodeResult{ .r = 0, .s = 0, .err = "null input | empty input" };
    }

    const b0 = b[0];

    if (b0 < 0x80) { // ASCII
        if (b.len > 1 and (b[1] & 0xC0) == 0x80) {
            return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };
        }
        return DecodeResult{ .r = b0, .s = 1, .err = null };
    } else if ((b0 & 0xE0) == 0xC0) { // 2 bytes
        if (b.len < 2) return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };
        if (b.len > 2 and (b[2] & 0xC0) == 0x80) return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };

        const b1 = b[1];
        if ((b1 & 0xC0) != 0x80) return DecodeResult{ .r = 0, .s = 0, .err = "invalid continuation byte" };

        const r: u32 = (@as(u32, b0 & 0x1F) << 6) | @as(u32, b1 & 0x3F);

        const invalid = validateRune(r, b0, b1, 2);
        if (invalid) |_| return invalid;

        return DecodeResult{ .r = r, .s = 2, .err = null };
    } else if ((b0 & 0xF0) == 0xE0) { // 3 bytes
        if (b.len < 3) return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };
        if (b.len > 3 and (b[3] & 0xC0) == 0x80) return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };

        const b1 = b[1];
        const b2 = b[2];

        if ((b1 & 0xC0) != 0x80 or (b2 & 0xC0) != 0x80)
            return DecodeResult{ .r = 0, .s = 0, .err = "invalid continuation byte" };

        const r: u32 = (@as(u32, b0 & 0x0F) << 12) |
            (@as(u32, b1 & 0x3F) << 6) |
            @as(u32, b2 & 0x3F);

        const invalid = validateRune(r, b0, b1, 3);
        if (invalid) |_| return invalid;

        return DecodeResult{ .r = r, .s = 3, .err = null };
    } else if ((b0 & 0xF8) == 0xF0) { // 4 bytes
        if (b.len < 4) return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };
        if (b.len > 4 and (b[4] & 0xC0) == 0x80) return DecodeResult{ .r = 0, .s = 0, .err = "invalid length" };

        const b1 = b[1];
        const b2 = b[2];
        const b3 = b[3];

        if ((b1 & 0xC0) != 0x80 or (b2 & 0xC0) != 0x80 or (b3 & 0xC0) != 0x80)
            return DecodeResult{ .r = 0, .s = 0, .err = "invalid continuation byte" };

        const r: u32 = (@as(u32, b0 & 0x07) << 18) |
            (@as(u32, b1 & 0x3F) << 12) |
            (@as(u32, b2 & 0x3F) << 6) |
            @as(u32, b3 & 0x3F);

        const invalid = validateRune(r, b0, b1, 4);
        if (invalid) |_| return invalid;

        return DecodeResult{ .r = r, .s = 4, .err = null };
    }

    return DecodeResult{ .r = 0, .s = 0, .err = "invalid utf8" };
}

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();

    const tests = [_][]const u8{ "A", "¢", "ह", "𐍈" };
    const names = [_][]const u8{ "Test1", "Test2", "Test3", "Test4" };

    for (tests, 0..) |input, i| {
        const result = decodeRune(input);
        try stdout.print("{s}: r={}, Rune: U+{X:0>4}, s={}, err={s}\n", .{ names[i], result.r, result.r, result.s, result.err orelse "NULL" });
    }
}

File: decode_utf-8/main.c
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include "core/core.h"

// Char. number range  |        UTF-8 octet sequence
//    (hexadecimal)    |              (binary)
// --------------------+---------------------------------------------
// 0000 0000-0000 007F | 0xxxxxxx
// 0000 0080-0000 07FF | 110xxxxx 10xxxxxx
// 0000 0800-0000 FFFF | 1110xxxx 10xxxxxx 10xxxxxx
// 0001 0000-0010 FFFF | 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx

DecodeResult decodeRune(const unsigned char* b, size_t len) {

    if (b == NULL || len <= 0) 
    { return (DecodeResult){ .r = 0, .s = 0, .err = "null input | empty input"};}
    
    unsigned char b0 = b[0];

    if (b0 < 0x80) { // ASCII
        
        if (len > 1) if ((b[1] & 0xC0) == 0x80) return (DecodeResult)
        { .r = 0, .s = 0, .err = "invalid length"};
        
        return (DecodeResult){ .r = (uint32_t)b0, .s = 1, .err = NULL};
    }
    else if ((b0 & 0xE0) == 0xC0) { // 2 byte character
        
        if (len < 2) return (DecodeResult){ .r = 0, .s = 0, .err = "invalid length"};
        if (len > 2) if ((b[2] & 0xC0) == 0x80) return (DecodeResult)
        { .r = 0, .s = 0, .err = "invalid length"};

        unsigned char b1 = b[1];

        if ((b1 & 0xC0) != 0x80) 
        {    return (DecodeResult){ .r = 0, .s = 0, .err = "invalid continuation byte"};}

        uint32_t r = ((uint32_t)(b0 & 0x1F) << 6) |
                    (uint32_t)(b1 & 0x3F);

        if (r < 0x80) return (DecodeResult){ .r = 0, .s = 0, .err = "overlong"};
        if (r >= 0xD800 && r <= 0xDFFF) return (DecodeResult){ .r = 0, .s = 0, .err = "surrogate halfs"};
        if (r > 0x10FFFF) return (DecodeResult){ .r = 0, .s = 0, .err = "too big"};
        
        return (DecodeResult){ .r = r, .s = 2, .err = NULL};
    }
    else if ((b0 & 0xF0) == 0xE0) { // 3 byte character
        
        if (len < 3) return (DecodeResult){ .r = 0, .s = 0, .err = "invalid length"};
        if (len > 3) if ((b[3] & 0xC0) == 0x80) return (DecodeResult)
        { .r = 0, .s = 0, .err = "invalid length"};
        
        unsigned char b1 = b[1]; unsigned char b2 = b[2];

        if ((b1 & 0xC0) != 0x80 || (b2 & 0xC0) != 0x80) 
        {   return (DecodeResult){ .r = 0, .s = 0, .err = "invalid continuation byte"};}
        
        uint32_t r = ((uint32_t)(b0 & 0x0F) << 12) |
                    ((uint32_t)(b1 & 0x3F) << 6) |
                    (uint32_t)(b2 & 0x3F);

        if (r < 0x800 || (b0 == 0xE0 && b1 < 0xA0)) return (DecodeResult){ .r = 0, .s = 0, .err = "overlong"};
        if (r >= 0xD800 && r <= 0xDFFF) return (DecodeResult){ .r = 0, .s = 0, .err = "surrogate halfs"};
        if (r > 0x10FFFF) return (DecodeResult){ .r = 0, .s = 0, .err = "too big"};
        
        return (DecodeResult){ .r = r, .s = 3, .err = NULL};
    }
    else if ((b0 & 0xF8) == 0xF0) { // 4 byte character
        
        if (len < 4) return (DecodeResult){ .r = 0, .s = 0, .err = "invalid length"};
        if (len > 4) if ((b[4] & 0xC0) == 0x80) return (DecodeResult)
        { .r = 0, .s = 0, .err = "invalid length"};
        
        unsigned char b1 = b[1]; unsigned char b2 = b[2]; unsigned char b3 = b[3];

        if ((b1 & 0xC0) != 0x80 || (b2 & 0xC0) != 0x80 || (b3 & 0xC0) != 0x80) 
        {    return (DecodeResult){ .r = 0, .s = 0, .err = "invalid continuation byte"};}

        uint32_t r = ((uint32_t)(b0 & 0x07) << 18) |
                    ((uint32_t)(b1 & 0x3F) << 12) |
                    ((uint32_t)(b2 & 0x3F) << 6) |
                    (uint32_t)(b3 & 0x3F);

        if (r < 0x10000 || (b0 == 0xF0 && b1 < 0x90)) return (DecodeResult){ .r = 0, .s = 0, .err = "overlong"};
        if (r >= 0xD800 && r <= 0xDFFF) return (DecodeResult){ .r = 0, .s = 0, .err = "surrogate halfs"};
        if (r > 0x10FFFF) return (DecodeResult){ .r = 0, .s = 0, .err = "too big"};
        
        return (DecodeResult){ .r = r, .s = 4, .err = NULL};
    }
    else return (DecodeResult){ .r = 0, .s = 0, .err = "invalid utf8"};
}

// Test function
int main(void) {
    // Test cases
    const char* test1 = "A";  // 1 byte
    const char* test2 = "¢";  // 2 bytes
    const char* test3 = "ह";  // 3 bytes
    const char* test4 = "𐍈";  // 4 bytes
    
    DecodeResult result;
    
    result = decodeRune((const unsigned char*)test1, strlen(test1));
    printf("Test1: r=%d, Rune: U+%04X, s=%d, err=%s\n", result.r, result.r,
        result.s, result.err ? result.err : "NULL");
    result = decodeRune((const unsigned char*)test2, strlen(test2));
    printf("Test2: r=%d, Rune: U+%04X, s=%d, err=%s\n", result.r, result.r,
        result.s, result.err ? result.err : "NULL");
    result = decodeRune((const unsigned char*)test3, strlen(test3));
    printf("Test3: r=%d, Rune: U+%04X, s=%d, err=%s\n", result.r, result.r, 
        result.s, result.err ? result.err : "NULL");   
    result = decodeRune((const unsigned char*)test4, strlen(test4));
    printf("Test4: r=%d, Rune: U+%04X, s=%d, err=%s\n", result.r, result.r, 
        result.s, result.err ? result.err : "NULL");
    
    return 0;
}